rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,416 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Client for Key Lease operations.
+
+use std::{sync::Arc, time::Duration};
+
+use tokio::{sync::Mutex, task::JoinHandle};
+
+use crate::leased_lock::{Error, ErrorKind, LeaseObservation, Response, SetCondition, SetOptions};
+use crate::state_store::{self};
+use azure_iot_operations_mqtt::interface::ManagedClient;
+use azure_iot_operations_protocol::common::hybrid_logical_clock::HybridLogicalClock;
+
+/// Key Lease client struct.
+pub struct Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    state_store: Arc<state_store::Client<C>>,
+    key_name: Vec<u8>,
+    lease_holder_name: Vec<u8>,
+    current_fencing_token: Arc<Mutex<Option<HybridLogicalClock>>>,
+    auto_renewal_task: JoinHandle<()>,
+}
+
+/// Key Lease client implementation
+///
+/// Notes:
+/// Do not call any of the methods of this client after the `state_store` parameter is shutdown.
+/// Calling any of the methods in this implementation after the `state_store` is shutdown results in undefined behavior.
+impl<C> Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync,
+    C::PubReceiver: Send + Sync,
+{
+    /// Create a new Key Lease Client.
+    ///
+    /// Notes:
+    /// - `lease_holder_name` is expected to be the client ID used in the underlying MQTT connection settings.
+    /// - There must be one instance of `leased_lock::Client` per lease.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`KeyLengthZero`](ErrorKind::KeyLengthZero) if the `key_name` is empty
+    ///
+    /// [`struct@Error`] of kind [`LeaseHolderNameLengthZero`](ErrorKind::LeaseHolderNameLengthZero) if the `lease_holder_name` is empty
+    pub fn new(
+        state_store: Arc<state_store::Client<C>>,
+        key_name: Vec<u8>,
```

Comment:
if a lease is over a key, should this be the key name or the lease name?

Reply:
Changed to lease_name, as per offline discussion.